### PR TITLE
Added event hook and track events list

### PR DESCRIPTION
**Pull request by JakeBat**
# Problem
Enterprise customers want to integrate SDK user activity into their own reporting system. Our current woopra integration is not fit for purpose because:
1) data is too woopra-specific 
2) raw data not exportable 
3) data is pseudonymised with no applicant ID association

# Solution
A new `UserAnalytics` event has been added to the `Tracker/index.js`. This new event has been added to the `sendEvent()` method in `Tracker/index.js` and will be dispatched when ever pre-existing analytic event within a specific array is called. Documentation for these changes will be added in ticket.

# Checklist
- [x] Has the CHANGELOG been updated?
- [x] Has the README been updated?
- [ ] Has the RELEASE_GUIDELINES been updated?
- [ ] Has the Sample App been updated?
- [ ] Has the MIGRATION doc been updated for any MAJOR breaking changes?
- [ ] Has the MIGRATION doc been updated for any MINOR breaking changes, including any translation strings or keys changes?
- [ ] Have any new automated tests been implemented or the existing ones changed?
- [ ] Have any new manual tests been written down or the existing ones changed?
- [ ] Have any new strings been translated or the existing ones changed?

### CODE REVIEW

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by stefaniacardenas** [comment]
Hey @JakeBat, just a reminder to avoid including ticket numbers, internal links and clients names in public repos. I'll edit the description for now.

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://920-pr-onfido-sdk-ui-onfido.surge.sh

File: src/Tracker/index.js [hunk 1]
```diff
@@ -82,9 +97,15 @@ const formatProperties = properties => {
   )
 }
 
+const userAnalyticsEvent = (eventName, properties) =>
+  dispatchEvent(new CustomEvent('userAnalyticsEvent', {detail: {eventName, properties}}));
+
 const sendEvent = (eventName, properties) => {
-  if (shouldSendEvents)
+  if (shouldSendEvents) {
+    if (integratorTrackedEvents.includes(eventName))
```
Comment (stefaniacardenas) [inline src/Tracker/index.js]: Minor: everywhere else in the code we enclose the consequent of an `if` statement into curly brackets, like so `if (condition) { consequent_logic }`. We will be adding stricter liner rules this sprint.

File: README.md [hunk 2]
```diff
@@ -542,6 +543,53 @@ Finally, as you are testing with the sandbox token, please be aware that the res
...
+
+### Overriding the hook
+
+In order to expose the user's progress through the SDK an `EventListener` must be added that listens for `UserAnalyticsEvent` events. This can be done anywhere within your application and might look something like the following:
+
+```javascript
+addEventListener('userAnalyticsEvent', (event) => //Your code here);
+```
+
+The code inside of the `EventListener`will now be called when a particular event is triggered, usually when the user reaches a new screen. For a full list of events see the bottom of this section.
```
Comment (seewah) [inline README.md]: missing space between `EventListener` and _will_

**Comment by seewah** [comment]
Thanks @JakeBat. however, as far as I can tell, this solution is probably not taking into account events happening while user is on the cross device journey. is that right? maybe @stefaniacardenas can provide more info?

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://920-pr-onfido-sdk-ui-onfido.surge.sh

File: src/Tracker/index.js [hunk 3]
```diff
@@ -14,6 +14,21 @@ let sentryHub= null
 
 const woopra = new WoopraTracker("onfidojssdkwoopra")
 
+const integratorTrackedEvents = [
+    'screen_welcome',
+    'screen_document_front_capture_file_upload',
+    'screen_document_front_confirmation',
```
[the remark this answers is not in the document]
Comment (JakeBat) [reply src/Tracker/index.js]: It was a conscious decision, there was a list provided that we wanted to expose. screen_document_type_select and screen_country_select are included in it but both were specified in the comments  to not be tracked. I'll send you over a link via slack shortly.

File: src/components/crossDevice/MobileFlow.js [hunk 4]
```diff
@@ -9,6 +9,8 @@ class MobileFlow extends Component {
     this.props.socket.on('disconnect ping', this.onDisconnectPing)
     this.props.socket.on('get config', this.sendConfig)
     this.props.socket.on('client success', this.onClientSuccess)
+
+    addEventListener('userAnalyticsEvent', (event) => this.sendUserAnalyticsEvent(event));
```
Comment (stefaniacardenas) [inline src/components/crossDevice/MobileFlow.js]: I left a very long message here but it got lost when I clicked on "Start review". Summary:
This component is part of the desktop flow, despite its unfortunate name, so this won't work. You would need to make sure that the "desktop" SDK is aware that the SDK running on id.onfido.com is calling `sendEvent`. Therefore you will have to do send a message from the sdk running on id.onfido.com (aka cross device client) to desktop by calling `socket.emit(roomId, 'user analytics event', event)` every time the `sendEvent` function is called and `mobileFlow` is `true` (`mobileFlow` is true when the SDK is running inside `id.onfido.com`). The message will have to be handled by the "desktop" SDK which will then call sendEvent again, this time firing the actual event. We can probably agree on a better design for this and if this sounds confusing we can have a call later.

Comment (stefaniacardenas) [inline src/components/crossDevice/MobileFlow.js]: I left a very long message here but it got lost when I clicked on "Start review". Summary:
This component is part of the desktop flow, despite its unfortunate name, so this won't work. You would need to make sure that the "desktop" SDK is aware that the SDK running on id.onfido.com is calling `sendEvent`. Therefore you will have to do send a message from the sdk running on id.onfido.com (aka cross device client) to desktop by calling `socket.emit(roomId, 'user analytics event', event)` every time the `sendEvent` function is called and `mobileFlow` is `true` (`mobileFlow` is true when the SDK is running inside `id.onfido.com`). The message will have to be handled by the "desktop" SDK which will then call sendEvent again, this time firing the actual event. We can probably agree on a better design for this and if this sounds confusing we can have a call later.

File: src/Tracker/index.js [hunk 5]
```diff
@@ -82,9 +97,16 @@ const formatProperties = properties => {
   )
 }
 
+const userAnalyticsEvent = (eventName, properties) =>
+  dispatchEvent(new CustomEvent('userAnalyticsEvent', {detail: {eventName, properties}}));
+
 const sendEvent = (eventName, properties) => {
-  if (shouldSendEvents)
+  if (shouldSendEvents) {
+    if (integratorTrackedEvents.includes(eventName)) {
```
[the remark this answers is not in the document]
Comment (JakeBat) [reply src/Tracker/index.js]: I haven't tried as I don't know how to get `eventName` and `properties` within `trackComponent`. If it's possible I have no issue moving it over.

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://920-pr-onfido-sdk-ui-onfido.surge.sh

File: README.md [hunk 6]
```diff
@@ -542,6 +543,53 @@ Finally, as you are testing with the sandbox token, please be aware that the res
...
+DOCUMENT_CAPTURE_CONFIRMATION_BACK - User reached the "document confirmation" screen for the back side (for two-sided document) 
+DOCUMENT_UPLOAD - User's document is uploading 
+FACIAL_INTRO - User reached the "selfie intro" screen 
+FACIAL_CAPTURE - User reached the "selfie capture" screen 
+FACIAL_CAPTURE_CONFIRMATION - User reached the "selfie confirmation" screen 
+FACIAL_UPLOAD - User's selfie is uploading 
+VIDEO_FACIAL_INTRO - User reached the "liveness intro" screen 
+VIDEO_FACIAL_CAPTURE - User reached the "liveness video capture" screen 
+VIDEO_FACIAL_CAPTURE_STEP_1 - User reached the 1st challenge during "liveness video capture", challenge_type can be found in eventProperties 
+VIDEO_FACIAL_CAPTURE_STEP_2 - User reached the 1st challenge during "liveness video capture", challenge_type can be found in eventProperties
```
Comment (eraxyz) [inline README.md]: 2nd?

Comment (JakeBat) [reply README.md]: Nice catch, thanks, updated.

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-920-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://920-pr-onfido-sdk-ui-onfido.surge.sh

**Pull request merged**